source/opt/licm_pass.cpp

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "opt/licm_pass.h"
+#include "opt/module.h"
+#include "opt/pass.h"
+
+#include <queue>
+#include <utility>
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status LICMPass::Process(ir::IRContext* context) {
+  InitializeProcessing(context);
+  bool modified = false;
+
+  if (context != nullptr) {
+    modified = ProcessIRContext();
+  }
+
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+bool LICMPass::ProcessIRContext() {
+  bool modified = false;
+  ir::Module* module = get_module();
+
+  // Process each function in the module
+  for (ir::Function& f : *module) {
+    modified |= ProcessFunction(&f);
+  }
+  return modified;
+}
+
+bool LICMPass::ProcessFunction(ir::Function* f) {
+  bool modified = false;
+  ir::LoopDescriptor* loop_descriptor = context()->GetLoopDescriptor(f);
+
+  // Process each loop in the function
+  for (ir::Loop& loop : *loop_descriptor) {
+    // Ignore nested loops, as we will process them in order in ProcessLoop
+    if (loop.IsNested()) {
+      continue;
+    }
+    modified |= ProcessLoop(&loop, f);
+  }
+  return modified;
+}
+
+bool LICMPass::ProcessLoop(ir::Loop* loop, ir::Function* f) {
+  bool modified = false;
+
+  // Process all nested loops first
+  for (ir::Loop* nested_loop : *loop) {
+    modified |= ProcessLoop(nested_loop, f);
+  }
+
+  // Get the pre_header. If it doesn't exist, force it's creation
+  ir::BasicBlock* pre_header_bb = loop->GetPreHeaderBlock();
+  if (!pre_header_bb) {
+    modified = true;
+    pre_header_bb = loop->GetOrCreatePreHeaderBlock();
+  }
+  ir::BasicBlock* header_bb = loop->GetHeaderBlock();
+  ir::BasicBlock* merge_bb = loop->GetMergeBlock();
+  std::queue<ir::BasicBlock*> loop_bbs{};
```

Comment:
Yes, we want to avoid std::deque and std::list, and std::queue is built on std::deque by default.  Please try to rework this.  Since you don't have to add too many basic block, you could do something like:

```
std::vector<BasicBlock*> loop_bbs;
...
for( int i = 0; i < loop_bbs.size(); ++i ) {
  ir::Basicblock* bb = loop_bbs[i];
  // do not delete the element.
  modified |= AnalyseAndHoistFromBB(loop, f, header_bb,
                                      merge_bb, pre_header_bb, bb, &loop_bbs);
}
```